File: domain/src/test/scala/net/kemitix/thorp/domain/RemoteKeyTest.scala
package net.kemitix.thorp.domain

import java.io.File
import java.nio.file.Paths

import org.scalatest.FreeSpec
import zio.{DefaultRuntime, UIO}

class RemoteKeyTest extends FreeSpec {

  private val emptyKey = RemoteKey.create("")

  "create a RemoteKey" - {
    "can resolve a path" - {
      "when key is empty" in {
        val key      = emptyKey
        val path     = "path"
        val expected = RemoteKey.create("path")
        val result   = key.resolve(path)
        assertResult(expected)(result)
      }
      "when path is empty" in {
        val key      = RemoteKey.create("key")
        val path     = ""
        val expected = RemoteKey.create("key")
        val result   = key.resolve(path)
        assertResult(expected)(result)
      }
      "when key and path are empty" in {
        val key      = emptyKey
        val path     = ""
        val expected = emptyKey
        val result   = key.resolve(path)
        assertResult(expected)(result)
      }
    }
    "asFile" - {
      "when key and prefix are non-empty" in {
        val key      = RemoteKey.create("prefix/key")
        val source   = Paths.get("source")
        val prefix   = RemoteKey.create("prefix")
        val expected = Some(new File("source/key"))
        val result   = key.asFile(source, prefix)
        assertResult(expected)(result)
      }
      "when prefix is empty" in {
        val key      = RemoteKey.create("key")
        val source   = Paths.get("source")
        val prefix   = emptyKey
        val expected = Some(new File("source/key"))
        val result   = key.asFile(source, prefix)
        assertResult(expected)(result)
      }
      "when key is empty" in {
        val key      = emptyKey
        val source   = Paths.get("source")
        val prefix   = RemoteKey.create("prefix")
        val expected = None
        val result   = key.asFile(source, prefix)
        assertResult(expected)(result)
      }
      "when key and prefix are empty" in {
        val key      = emptyKey
        val source   = Paths.get("source")
        val prefix   = emptyKey
        val expected = None
        val result   = key.asFile(source, prefix)
        assertResult(expected)(result)
      }
    }
    "fromSourcePath" - {
      "when path in source" in {
        val source   = Paths.get("/source")
        val path     = source.resolve("/source/child")
        val expected = RemoteKey.create("child")
        val result   = RemoteKey.fromSourcePath(source, path)
        assertResult(expected)(result)
      }
    }
    "from source, prefix, file" - {
      "when file in source" in {
        val source   = Paths.get("/source")
        val prefix   = RemoteKey.create("prefix")
        val file     = new File("/source/dir/filename")
        val expected = RemoteKey.create("prefix/dir/filename")
        val program  = UIO(RemoteKey.from(source, prefix, file))
        val result   = new DefaultRuntime {}.unsafeRunSync(program).toEither
        assertResult(Right(expected))(result)
      }
    }
  }
  "asFile" - {
    "remoteKey is empty" in {
      val source    = Paths.get("/source")
      val prefix    = RemoteKey.create("prefix")
      val remoteKey = RemoteKey.create("")

      val expected = None

      val result = remoteKey.asFile(source, prefix)

      assertResult(expected)(result)
    }
    "remoteKey is not empty" - {
      "remoteKey is within prefix" in {
        val source    = Paths.get("/source")
        val prefix    = RemoteKey.create("prefix")
        val remoteKey = RemoteKey.create("prefix/key")

        val expected = Some(Paths.get("/source/key").toFile)

        val result = remoteKey.asFile(source, prefix)

        assertResult(expected)(result)
      }
      "remoteKey is outwith prefix" in {
        val source    = Paths.get("/source")
        val prefix    = RemoteKey.create("prefix")
        val remoteKey = RemoteKey.create("elsewhere/key")

        val expected = None

        val result = remoteKey.asFile(source, prefix)

        assertResult(expected)(result)
      }
    }
  }

}

File: domain/src/test/java/net/kemitix/thorp/domain/RemoteKeyTest.java
package net.kemitix.thorp.domain;

import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

public class RemoteKeyTest
        implements WithAssertions {

    private RemoteKey emptyKey = RemoteKey.create("");

    @Nested
    @DisplayName("Create a RemoteKey")
    public class CreateRemoteKey {
        @Nested
        @DisplayName("resolve()")
        public class ResolvePath {
            @Test
            @DisplayName("key is empty")
            public void keyIsEmpty() {
                //given
                RemoteKey expected = RemoteKey.create("path");
                RemoteKey key = emptyKey;
                //when
                RemoteKey result = key.resolve("path");
                //then
                assertThat(result).isEqualTo(expected);
            }
            @Test
            @DisplayName("path is empty")
            public void pathIsEmpty() {
                //given
                RemoteKey expected = RemoteKey.create("key");
                RemoteKey key = RemoteKey.create("key");
                String path = "";
                //when
                RemoteKey result = key.resolve(path);
                //then
                assertThat(result).isEqualTo(expected);
            }
            @Test
            @DisplayName("key and path are empty")
            public void keyAndPathEmpty() {
                //given
                RemoteKey expected = RemoteKeyTest.this.emptyKey;
                String path = "";
                RemoteKey key = emptyKey;
                //when
                RemoteKey result = key.resolve(path);
                //then
                assertThat(result).isEqualTo(expected);
            }
        }
        @Nested
        @DisplayName("asFile()")
        public class AsFile {
            @Test
            @DisplayName("key and prefix are non-empty")
            public void keyAndPrefixNonEmpty() {
                //given
                Optional<File> expected = Optional.of(new File("source/key"));
                RemoteKey key = RemoteKey.create("prefix/key");
                Path source = Paths.get("source");
                RemoteKey prefix = RemoteKey.create("prefix");
                //when
                Optional<File> result = key.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }
            @Test
            @DisplayName("prefix is empty")
            public void prefixEmpty() {
                //given
                Optional<File> expected = Optional.of(new File("source/key"));
                RemoteKey key = RemoteKey.create("key");
                Path source = Paths.get("source");
                RemoteKey prefix = emptyKey;
                //when
                Optional<File> result = key.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }

            @Test
            @DisplayName("key is empty")
            public void keyEmpty() {
                //given
                Optional<File> expected = Optional.empty();
                RemoteKey key = emptyKey;
                Path source = Paths.get("source");
                RemoteKey prefix = RemoteKey.create("source/key");
                //when
                Optional<File> result = key.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }

            @Test
            @DisplayName("key and prefix are empty")
            public void keyAndPrefixEmpty() {
                //given
                Optional<File> expected = Optional.empty();
                RemoteKey key = emptyKey;
                Path source = Paths.get("source");
                RemoteKey prefix = emptyKey;
                //when
                Optional<File> result = key.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }
        }
        @Nested
        @DisplayName("fromSourcePath()")
        public class FromSourcePath {
            @Test
            @DisplayName("path is in source")
            public void pathInSource() {
                //given
                RemoteKey expected = RemoteKey.create("child");
                Path source = Paths.get("/source");
                Path path = source.resolve("/source/child");
                //when
                RemoteKey result = RemoteKey.fromSourcePath(source, path);
                //then
                assertThat(result).isEqualTo(expected);
            }
        }
        @Nested
        @DisplayName("from(source, prefix, file)")
        public class FromSourcePrefixFile {
            @Test
            @DisplayName("file in source")
            public void fileInSource() {
                //given
                RemoteKey expected = RemoteKey.create("prefix/dir/filename");
                Path source = Paths.get("/source");
                RemoteKey prefix = RemoteKey.create("prefix");
                File file = new File("/source/dir/filename");
                //when
                RemoteKey result = RemoteKey.from(source, prefix, file);
                //then
                assertThat(result).isEqualTo(expected);
            }
        }
    }
    @Nested
    @DisplayName("asFile()")
    public class AsFile {
        @Test
        @DisplayName("remoteKey is empty")
        public void remoteKeyEmpty() {
            //given
            Optional<File> expected = Optional.empty();
            Path source = Paths.get("/source");
            RemoteKey prefix = RemoteKey.create("prefix");
            RemoteKey remoteKey = emptyKey;
            //when
            Optional<File> result = remoteKey.asFile(source, prefix);
            //then
            assertThat(result).isEqualTo(expected);
        }
        @Nested
        @DisplayName("remoteKey is not empty")
        public class RemoteKeyNotEmpty {
            @Test
            @DisplayName("remoteKey is within prefix")
            public void remoteKeyWithinPrefix() {
                //given
                Optional<File> expected = Optional.of(new File("/source/key"));
                Path source = Paths.get("/source");
                RemoteKey prefix = RemoteKey.create("prefix");
                RemoteKey remoteKey = RemoteKey.create("prefix/key");
                //when
                Optional<File> result = remoteKey.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }

            @Test
            @DisplayName("remoteKey is outwith prefix")
            public void remoteKeyIsOutwithPrefix() {
                //given
                Optional<File> expected = Optional.empty();
                Path source = Paths.get("/source");
                RemoteKey prefix = RemoteKey.create("prefix");
                RemoteKey remoteKey = RemoteKey.create("elsewhere/key");
                //when
                Optional<File> result = remoteKey.asFile(source, prefix);
                //then
                assertThat(result).isEqualTo(expected);
            }
        }
    }

}
